clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1910,7 +1931,8 @@ void Sema::MarkDevice(void) {
   // Iterate through SYCL_EXTERNAL functions and add them to the device decls.
   for (const auto &entry : *Marker.SYCLCG.getRoot()) {
     if (auto *FD = dyn_cast<FunctionDecl>(entry.Callee->getDecl())) {
-      if (FD->hasAttr<SYCLDeviceAttr>() && !FD->hasAttr<SYCLKernelAttr>())
+      if (FD->hasAttr<SYCLDeviceAttr>() && !FD->hasAttr<SYCLKernelAttr>() &&
+          FD->hasBody())
```

Comment:
SYCL_EXTERNAL functions can be undefined right?

> In the case of linking C++ functions and methods to a SYCL application, where the definitions and declarations are not available in the same translation unit of the compiler, then the macro SYCL_EXTERNAL has to be provided. 

So I do not think this is correct. @premanandrao can you confirm?

Reply:
Yes, I think this will not consider declarations that are not definitions.

Reply:
@elizabethandrews , @premanandrao - can you please elaborate what exactly is incorrect?
hasBody() check is added not to process SYCL_EXTERAL declarations in the `for (Decl *D : syclDeviceDecls())` loop below.

Reply:
If a function is declared but not defined, and it is called from sycl device code, it won't get added in the device decls list.
Are you deliberately choosing not to add those to the list?

Reply:
yes. This is rather "device defs" by its purpose - the comment says
'  // We store SYCL Kernels here and handle separately -- which is a hack.
  // FIXME: It would be best to refactor this.'
SYCL_EXTERNAL functions with bodies are treated like wise - they are roots of the call graph too.

Reply:
Okay, thanks for the explanation.